datalad_next/runners/run.py

```diff
@@ -0,0 +1,166 @@
+    and logic inside the timeout handler that can trigger a termination or
+    kill signal to the subprocess, if the termination time or kill time has
+    come.
+
+    Parameters
+    ----------
+    cls : type[Protocol]
+        A protocol class that should be extended by the kill-interface
+
+    Returns
+    -------
+    Any
+        If the `cls` is not a subclass of `GeneratorMixIn`, the result of
+        `cls._prepare_result` will be returned.
+
+    Generator
+        If the protocol is a subclass of `GeneratorMixIn`, the `Generator` that
+        is a result of `ThreadedRunner.run()` will be returned.
+    """
+    class KillWrapper(cls):
+        def __init__(self, *args, **kwargs):
+            self.armed = kwargs.pop('armed')
+            self.introduced_timeout = kwargs.pop('introduced_timeout')
+            self.terminate_time = kwargs.pop('terminate_time')
+            kill_time = kwargs.pop('kill_time')
+            self.kill_time = (
+                ((self.terminate_time or 0) + kill_time)
+                if kill_time is not None
+                else kill_time
+            )
+            self.process: subprocess.Popen | None = None
+            self.kill_counter: int = 0
+            super().__init__(*args, **kwargs)
+
+        def arm(self) -> None:
+            self.kill_counter = 0
+            self.armed = True
+
+        def connection_made(self, process: subprocess.Popen) -> None:
+            self.process = process
+            super().connection_made(process)
+
+        def timeout(self, fd: int | None) -> bool:
+            if self.armed:
+                if self.kill_time and self.kill_counter >= self.kill_time:
+                    self.process.kill()
+                    self.kill_time = None
+                if self.terminate_time and self.kill_counter >= self.terminate_time:
+                    self.process.terminate()
+                    self.terminate_time = None
+                self.kill_counter += 1
+
+            # If we set the timeout argument due to a not None kill_time
+            # or a not-None terminate_time, and due to a None timeout parameter,
+            # we leave the timeout handler here.
+            if self.introduced_timeout:
+                return False
+
+            # If the timeout was set by the user of the context, we execute
+            # the timeout handler of the superclass.
+            return super().timeout(fd)
+
+    return KillWrapper
+
+
+@contextmanager
+def run(
+        cmd: list,
+        protocol_class: type[Protocol],
+        *,
+        cwd: Path | None = None,
+        input: int | IO | bytes | Queue[bytes | None] | None = None,
+        timeout: float | None = None,
+        terminate_time: int | None = None,
+        kill_time: int | None = None,
+) -> Any | Generator:
+
+    introduces_timeout = False
+    if timeout is None:
+        introduces_timeout = True
+        timeout = 1.0
```

Comment:
What is the logic here? Unconditional timeouts? Would this even work with any long-running process?

Reply:
The kill and termination mechanism is implemented largely in the `timeout`-callback. If the user does not specify timeouts, we introduce one second timeout to ensure that the `timeout`-callback is called. We also record the fact that we introduced the timeout in order to do the right thing at the end of our `timeout`-handler